### PR TITLE
Add valijson package

### PR DESCRIPTION
**Pull request by mathbunnyru**
Specify library name and version:  **valijson/0.3.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/valijson/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        include_folder = os.path.join(self._source_subfolder, "include")
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy(pattern="*", dst="include", src=include_folder)
+
+    def package_id(self):
+        self.info.header_only()
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "valijson"
+        self.cpp_info.names["cmake_find_package_multi"] = "valijson"
```
Comment (prince-chrismc) [inline recipes/valijson/all/conanfile.py:34]: This does not look like it coversthe CMake targets

https://github.com/tristanpenman/valijson#how-to-add-this-library-to-your-cmake-target

> target_link_libraries(your-executable ValiJSON::valijson)

Comment (mathbunnyru) [reply recipes/valijson/all/conanfile.py:34]: Oh, my wrong.
Could you please help me, how should I change it?

File: recipes/valijson/all/test_package/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    def requirements(self):
+        self.requires("rapidjson/cci.20200410")
+
+    def test(self):
+        if not tools.cross_building(self.settings):
+            bin_path = os.path.join("bin", "test_package")
+            schema_file = os.path.abspath(os.path.join(self.source_folder, "schema.json"))
+            valid_file = os.path.abspath(os.path.join(self.source_folder, "valid.json"))
+            invalid_file = os.path.abspath(os.path.join(self.source_folder, "invalid.json"))
```
Comment (prince-chrismc) [inline recipes/valijson/all/test_package/conanfile.py:22]: Could this be embedded into the cpp?

File: recipes/valijson/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+set(CMAKE_CXX_STANDARD 11)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (prince-chrismc) [inline recipes/valijson/all/test_package/CMakeLists.txt:9]: We should be using the official cmake target

File: recipes/valijson/all/test_package/conanfile.py (hunk 2 above)
Comment (mathbunnyru) [reply recipes/valijson/all/test_package/conanfile.py:22]: I tried to create an example as close to the usage section of valijson, as possible.
They do expect files to be passed to parse jsons from.
That's why I'm passing it here like this.

File: recipes/valijson/all/conanfile.py (hunk 1 above)
Comment (prince-chrismc) [reply recipes/valijson/all/conanfile.py:34]: Space is our expert, here's an example that recently got merged, https://github.com/conan-io/conan-center-index/pull/2559/files#diff-559d1628b195b97e815856556814668fR210-R226

If you need more help, ill do my best

File: recipes/valijson/all/test_package/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/valijson/all/test_package/conanfile.py:22]: That's fair, I prefer to having less extra support files. This is just a suggestion :smile:

File: recipes/valijson/all/conanfile.py (hunk 1 above)
Comment (uilianries) [reply recipes/valijson/all/conanfile.py:34]: by default it uses the package name, so you don't need it.

Comment (mathbunnyru) [reply recipes/valijson/all/conanfile.py:34]: Sorry, @prince-chrismc and @uilianries .
I don't understand how to fix the issue.

Should I rename my package?
Or should I somehow add components? There's only one component, I suppose, but I don't understand where ValiJSON::valijson should go.

Comment (prince-chrismc) [reply recipes/valijson/all/conanfile.py:34]: This should do the trick

```suggestion
    def package_info(self):
        
        # self.cpp_info.filenames["cmake_find_package"] = "valijson" # TBA: There's no installed config file
        # self.cpp_info.filenames["cmake_find_package_multi"] = "valijson" # TBA: There's no installed config file
        self.cpp_info.names["cmake_find_package"] = "ValiJSON"
        self.cpp_info.names["cmake_find_package_multi"] = "ValiJSON"

        self.cpp_info.components["libvalijson"].names["cmake_find_package"] = "valijson"
        self.cpp_info.components["libvalijson"].names["cmake_find_package_multi"] = "valijson"
        self.cpp_info.components["libvalijson"].libs = [self._lib_name("valijson")]
```

Comment (mathbunnyru) [reply recipes/valijson/all/conanfile.py]: "libvalijson" - is it some semi-random string? For example, what will change, if I name it lib_valijson or something else?

self._lib_name is missing, I think in this case it's just valijson

**Comment by mathbunnyru** [comment]
I think I made it work, but I'm not sure if I'm correct at all.

For example, I deleted `self.cpp_info.components["libvalijson"].libs = ["valijson"]`

File: recipes/valijson/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+set(CMAKE_CXX_STANDARD 11)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+find_package(valijson REQUIRED)
```
Comment (prince-chrismc) [inline recipes/valijson/all/test_package/CMakeLists.txt:8]:
```suggestion
find_package(ValiJSON REQUIRED)
```

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Comment by madebr** [comment]
@mathbunnyru 
valjson provides some additional libraries.
Aren't these needed? As an option?
https://github.com/tristanpenman/valijson/blob/master/CMakeLists.txt#L41-L60

**Comment by prince-chrismc** [comment]
> @mathbunnyru
> valjson provides some additional libraries.
> Aren't these needed? As an option?
> https://github.com/tristanpenman/valijson/blob/master/CMakeLists.txt#L41-L60

The consumer can specify which ever they want 🤔

**Comment by mathbunnyru** [comment]
@prince-chrismc is right.
The consumer might use json parser of their choice - jsoncpp, rapidjson, ....
It doesn't affect this package.

Actually, an example can be seen in the test_package - in this situation rapidjson is used and it is added as a dependency of test_package.

**Comment by madebr** [comment]
Thanks. It wasn't clear to me at first sight whether these libraries were part of the valijson library or external.

**Review by madebr — APPROVED** [review]

**Pull request merged**